### PR TITLE
feat(engine): integrate sparse trie into the state root task

### PR DESCRIPTION
**Pull request by shekhirin**
Replaces the `HashBuilder` proof calculation in state root task with the `SparseTrie`.

### CODE REVIEW

File: crates/engine/tree/src/tree/root.rs [hunk 1]
```diff
@@ -287,12 +278,17 @@ where
             };
 
             // TODO: replace with parallel proof
-            let result =
-                Proof::overlay_multiproof(provider.tx_ref(), input.as_ref().clone(), targets);
+            let result = Proof::overlay_multiproof(
+                provider.tx_ref(),
+                // TODO(alexey): this clone can be expensive, we should avoid it
+                input.as_ref().clone(),
+                targets.clone(),
```
Comment (rkrasiuk) [inline crates/engine/tree/src/tree/root.rs:285]: redundant clone?

File: crates/engine/tree/src/tree/root.rs [hunk 2]
```diff
@@ -334,28 +333,20 @@ where
             "Spawning root calculation"
         );
 
-        let tx = self.tx.clone();
-        let view = self.config.consistent_view.clone();
-        let input = self.config.input.clone();
-        let state = self.state.clone();
+        // TODO(alexey): store proof targets in `ProofSequecner` to avoid recomputing them
```
Comment (rkrasiuk) [inline crates/engine/tree/src/tree/root.rs:336]: 👍

File: crates/engine/tree/src/tree/root.rs (hunk 1 above)
Comment (shekhirin) [reply crates/engine/tree/src/tree/root.rs]: fixed

File: crates/engine/tree/src/tree/root.rs [hunk 3]
```diff
@@ -454,7 +449,10 @@ where
                                 roots_calculated,
                                 "All proofs processed, ending calculation"
                             );
-                            return Ok((current_root, trie_updates));
+                            let mut trie = self.sparse_trie.take().unwrap();
```
Comment (fgimenez) [inline crates/engine/tree/src/tree/root.rs:452]:
```suggestion
                            let mut trie = self.sparse_trie.take().expect("update is in progress and sparse trie should not be None");
```

File: crates/engine/tree/src/tree/root.rs [hunk 4]
```diff
@@ -454,7 +449,10 @@ where
...
+                            let root = trie.root().unwrap();
```
Comment (fgimenez) [inline crates/engine/tree/src/tree/root.rs:453]: same as above for these unwraps, not sure what the expect message should be here though

**Review by fgimenez — APPROVED** [review]
lgtm, nits about using `expect` instead of `unwrap`, I guess one of them is unavoidable here?

**Pull request merged**